server/market/market.go

```diff
@@ -194,81 +258,18 @@ func (m *Market) Cancelable(oid order.OrderID) bool {
 	return false
 }
 
-// NewMarket creates a new Market for the provided base and quote assets, with
-// an epoch cycling at given duration in milliseconds.
-func NewMarket(ctx context.Context, mktInfo *dex.MarketInfo, storage db.DEXArchivist,
-	swapper Swapper, authMgr AuthManager, coinLockerBase, coinLockerQuote coinlock.CoinLocker) (*Market, error) {
-	// Make sure the DEXArchivist is healthy before taking orders.
-	if err := storage.LastErr(); err != nil {
-		return nil, err
-	}
-
-	// Lock coins backing active orders.
-	baseCoins, quoteCoins, err := storage.ActiveOrderCoins(mktInfo.Base, mktInfo.Quote)
-	if err != nil {
-		return nil, err
-	}
-	// TODO: first check that the coins aren't already locked by e.g. another market.
-	coinLockerBase.LockCoins(baseCoins)
-	coinLockerQuote.LockCoins(quoteCoins)
-
-	// Supplement the provided context with a cancel function. The Stop method
-	// may be used to stop the market, or the parent context itself may be
-	// canceled.
-	ctxInternal, cancel := context.WithCancel(ctx)
-
-	return &Market{
-		ctx:             ctxInternal,
-		cancel:          cancel,
-		running:         make(chan struct{}),
-		marketInfo:      mktInfo,
-		epochDuration:   int64(mktInfo.EpochDuration),
-		orderRouter:     make(chan *orderUpdateSignal, 16),
-		book:            book.New(mktInfo.LotSize),
-		matcher:         matcher.New(),
-		epochOrders:     make(map[order.OrderID]order.Order),
-		swapper:         swapper,
-		auth:            authMgr,
-		storage:         storage,
-		coinLockerBase:  coinLockerBase,
-		coinLockerQuote: coinLockerQuote,
-	}, nil
-}
-
-// Start beings order processing.
-func (m *Market) Start(startEpochIdx int64) {
-	m.wg.Add(1)
-	go m.runEpochs(startEpochIdx)
-}
-
-// Stop begins the Market's shutdown. Use WaitForShutdown after Stop to wait for
-// shutdown to complete. The Swapper is NOT stopped.
-func (m *Market) Stop() {
-	log.Infof("Market shutting down...")
-	m.cancel()
-	for _, s := range m.orderFeeds {
-		close(s)
-	}
-}
-
-// WaitForShutdown waits until the order processing is finished and the Market
-// is stopped.
-func (m *Market) WaitForShutdown() {
-	m.wg.Wait()
-}
-
-// WaitForEpochOpen waits until the start of epoch processing.
-func (m *Market) WaitForEpochOpen() {
-	<-m.running
-}
-
-func (m *Market) notifyNewOrder(order order.Order, action bookUpdateAction) {
+func (m *Market) notifyNewOrder(ctx context.Context, order order.Order, action bookUpdateAction, eidx int64) {
 	// send to each receiver
+	m.orderFeedMtx.RLock()
+	defer m.orderFeedMtx.RUnlock()
+	if ctx.Err() != nil {
+		return
+	}
```

Comment:
Definitely is, although `m.orderFeedMtx.RLock()` could be blocking in `OrderFeed` or `Run`'s deferred function that closes all the order feeds. Since `processOrder` runs `notifyNewOrder` async, and `OrderFeed` is called by other goroutines, I don't think we can be sure the context has canceled by the time the mutext is acquired.

Reply:
That makes sense.